activerecord/test/cases/associations/belongs_to_associations_test.rb

```diff
@@ -38,10 +39,12 @@ def test_belongs_to
   end
 
   def test_assigning_belongs_to_on_destroyed_object
-    client = Client.create!(name: "Client")
+    firm = Firm.create!(name: "Firm")
+    client = Client.create!(name: "Client", firm: firm)
     client.destroy!
     assert_raise(frozen_error_class) { client.firm = nil }
-    assert_raise(frozen_error_class) { client.firm = Firm.new(name: "Firm") }
+    assert_raise(frozen_error_class) { client.firm = Firm.new(name: "New Firm") }
```

Comment:
If this change is required, it means that this PR will change existing public behavior in exchange for preloading on destroyed record.

Reply:
Yup. The change in public behavior is that frozen errors will only be raised if the foreign key is actually changing (except for the nil case that we punted on). This behavior is also inconsistent with changing "regular" attributes on destroyed models which will always throw an exception. Thinking through it a bit more I can see a few ways forward:

1. Make the assignments to any attributes (i.e. foreign keys or primitives) a no-op if they don't actually change anything but a) that doesn't feel particular useful outside of the preloader use case; b) the required additional equality checks could have a negative performance impact.
1. Rollback #31575 which caused the change in preloader behavior introduced in 5.2.0. That issue seems more serious than the preloader not working on destroyed models so I'm not crazy about rolling it back.
1. Define a separate API on `Association` for setting the `target` that would have the behavior needed by the preloader (i.e. it should never change the foreign key).
1. Punt on making the preloader work with destroyed models
1. Move forward with the change as currently implemented which gets preloading working on destroyed models but is inconsistent with setting attributes on destroyed models

Option 3 seems like the best option but I'm not sure it's worth it given Rails 5.2 has already been released with the change in preloader behavior and I've released a workaround in goldiloader. What are your thoughts?

Reply:
The behaviour of option 3 sounds correct, but the design feels wrong; we're adding a method that duplicates the implementation of `target=` so that we can avoid the behaviour added in `BelongsToAssociation`, in order to fix a bug that only affects belongs to associations.

Another potential solution is to revert https://github.com/rails/rails/pull/31575 and https://github.com/rails/rails/pull/32338, and solve the original problem in a different way: set both the target _and_ the foreign key when assigning an inverse association.

Reply:
Right, my preference is here:

```diff
diff --git a/activerecord/lib/active_record/associations/association.rb b/activerecord/lib/active_record/associations/association.rb
index d883eda52d..9301ccc124 100644
--- a/activerecord/lib/active_record/associations/association.rb
+++ b/activerecord/lib/active_record/associations/association.rb
@@ -18,8 +18,7 @@ module Associations
     #       HasManyAssociation + ForeignAssociation
     #         HasManyThroughAssociation + ThroughAssociation
     class Association #:nodoc:
-      attr_reader :owner, :target, :reflection
-      attr_accessor :inversed
+      attr_reader :owner, :target, :reflection, :inversed
 
       delegate :options, to: :reflection
 
@@ -76,11 +75,6 @@ def target=(target)
         loaded!
       end
 
-      def preloaded_target=(target) #:nodoc:
-        @target = target
-        loaded!
-      end
-
       def scope
         target_scope.merge!(association_scope)
       end
@@ -103,23 +97,24 @@ def reset_scope
 
       # Set the inverse association, if possible
       def set_inverse_instance(record)
-        if invertible_for?(record)
-          inverse = record.association(inverse_reflection_for(record).name)
-          inverse.target = owner
-          inverse.inversed = true
+        if inverse = inverse_association_for(record)
+          inverse.inversed_from(owner)
         end
         record
       end
 
       # Remove the inverse association, if possible
       def remove_inverse_instance(record)
-        if invertible_for?(record)
-          inverse = record.association(inverse_reflection_for(record).name)
-          inverse.target = nil
-          inverse.inversed = false
+        if inverse = inverse_association_for(record)
+          inverse.inversed_from(nil)
         end
       end
 
+      def inversed_from(record)
+        self.target = record
+        @inversed = !!record
+      end
+
       # Returns the class of the target. belongs_to polymorphic overrides this to look at the
       # polymorphic_type field on the owner.
       def klass
@@ -245,6 +240,12 @@ def raise_on_type_mismatch!(record)
           end
         end
 
+        def inverse_association_for(record)
+          if invertible_for?(record)
+            record.association(inverse_reflection_for(record).name)
+          end
+        end
+
         # Can be redefined by subclasses, notably polymorphic belongs_to
         # The record parameter is necessary to support polymorphic inverses as we must check for
         # the association in the specific class of the record.
diff --git a/activerecord/lib/active_record/associations/belongs_to_association.rb b/activerecord/lib/active_record/associations/belongs_to_association.rb
index 1109fee462..e5fcdb06dc 100644
--- a/activerecord/lib/active_record/associations/belongs_to_association.rb
+++ b/activerecord/lib/active_record/associations/belongs_to_association.rb
@@ -26,10 +26,12 @@ def replace(record)
           decrement_counters
         end
 
+        replace_keys(record)
+
         self.target = record
       end
 
-      def target=(record)
+      def inversed_from(record)
         replace_keys(record)
         super
       end
diff --git a/activerecord/lib/active_record/associations/preloader/association.rb b/activerecord/lib/active_record/associations/preloader/association.rb
index f076f53af0..d6f7359055 100644
--- a/activerecord/lib/active_record/associations/preloader/association.rb
+++ b/activerecord/lib/active_record/associations/preloader/association.rb
@@ -46,7 +46,7 @@ def associate_records_to_owner(owner, records)
             if reflection.collection?
               association.target.concat(records)
             else
-              association.preloaded_target = records.first unless records.empty?
+              association.target = records.first unless records.empty?
             end
           end
 
```